python/ray/air/preprocessors/stateful_preprocessor.py

```diff
@@ -0,0 +1,120 @@
+from typing import Callable, TYPE_CHECKING, Dict
+
+from ray.air.preprocessor import Preprocessor
+from ray.data import Dataset
+
+if TYPE_CHECKING:
+    import pandas
+
+
+class StatefulPreprocessor(Preprocessor):
+    """Implements a stateful preprocessor that fits on a Dataset.
+
+    This is meant to be generic and can be used to perform arbitrary stateful
+    preprocessing that cannot already be done through existing preprocessors.
+    Logic must be defined to perform fitting on a Ray Dataset and transforming
+    Pandas DataFrames.
```

Comment:
```suggestion
    pandas DataFrames.
```
(pandas uses lower-case for its name)